// testlink-xmlrpc/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportanceLevel = exports.StepAction = exports.TestCaseStatus = exports.Details = exports.ExecutionType = exports.ExecutionStatus = exports.ActionOnDuplicate = void 0;
var ActionOnDuplicate;
(function (ActionOnDuplicate) {
    ActionOnDuplicate["BLOCK"] = "block";
    ActionOnDuplicate["GENERATE_NEW"] = "generate_new";
    ActionOnDuplicate["CREATE_NEW_VERSION"] = "create_new_version";
})(ActionOnDuplicate = exports.ActionOnDuplicate || (exports.ActionOnDuplicate = {}));
var ExecutionStatus;
(function (ExecutionStatus) {
    ExecutionStatus["NOT_RUN"] = "n";
    ExecutionStatus["PASSED"] = "p";
    ExecutionStatus["FAILED"] = "f";
    ExecutionStatus["BLOCKED"] = "b";
})(ExecutionStatus = exports.ExecutionStatus || (exports.ExecutionStatus = {}));
var ExecutionType;
(function (ExecutionType) {
    ExecutionType[ExecutionType["MANUAL"] = 1] = "MANUAL";
    ExecutionType[ExecutionType["AUTOMATED"] = 2] = "AUTOMATED";
})(ExecutionType = exports.ExecutionType || (exports.ExecutionType = {}));
var Details;
(function (Details) {
    Details["FULL"] = "full";
    Details["SIMPLE"] = "simple";
    Details["VALUE"] = "value";
})(Details = exports.Details || (exports.Details = {}));
var TestCaseStatus;
(function (TestCaseStatus) {
    TestCaseStatus[TestCaseStatus["FINAL"] = 7] = "FINAL";
    TestCaseStatus[TestCaseStatus["FUTURE"] = 6] = "FUTURE";
    TestCaseStatus[TestCaseStatus["OBSOLETE"] = 5] = "OBSOLETE";
    TestCaseStatus[TestCaseStatus["REWORK"] = 4] = "REWORK";
    TestCaseStatus[TestCaseStatus["REVIEW_IN_PROGRESS"] = 3] = "REVIEW_IN_PROGRESS";
    TestCaseStatus[TestCaseStatus["READY_FOR_REVIEW"] = 2] = "READY_FOR_REVIEW";
    TestCaseStatus[TestCaseStatus["DRAFT"] = 1] = "DRAFT";
})(TestCaseStatus = exports.TestCaseStatus || (exports.TestCaseStatus = {}));
var StepAction;
(function (StepAction) {
    StepAction["CREATE"] = "create";
    StepAction["UPDATE"] = "update";
    StepAction["PUSH"] = "push"; // Shift down all steps with step number >= step number provided (NOT IMPLEMENTED YET ON TESTLINK)
})(StepAction = exports.StepAction || (exports.StepAction = {}));
var ImportanceLevel;
(function (ImportanceLevel) {
    ImportanceLevel[ImportanceLevel["HIGH"] = 3] = "HIGH";
    ImportanceLevel[ImportanceLevel["MEDIUM"] = 2] = "MEDIUM";
    ImportanceLevel[ImportanceLevel["LOW"] = 1] = "LOW";
})(ImportanceLevel = exports.ImportanceLevel || (exports.ImportanceLevel = {}));
